### PR TITLE
[DPE-3634] Rework spark-history-server into new pattern

### PR DESCRIPTION
**Pull request by deusebio**
Refactoring of Spark History server according to new pattern

@marcoppenheimer @zmraul

### CODE REVIEW

File: lib/charms/data_platform_libs/v0/data_interfaces.py
Comment (deusebio) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:1]: Using a new version I have taken from the ZooKeeper PR, to have a data-interface class with separation of concerns between databag parsing and handlers.

File: src/workload.py [hunk 1]
```diff
@@ -3,199 +3,83 @@
...
+        """Return current environment."""
+        if self._envs is not None:
+            return self._envs
+
+        self._envs = self.from_env(self.read(self.ENV_FILE)) if self.exists(self.ENV_FILE) else {}
 
-    def get_truststore_file(self, mode: IOMode) -> ContainerFile:
-        """Return the configuration file for Spark History server."""
-        return ContainerFile(self.container, self.user, self.SPARK_TRUSTSTORE, mode)
+        return self.envs
```
[the remark this answers is not in the document]
Comment (deusebio) [reply src/workload.py:44]: Yes, the function is just re-entering the property but at this point `self._envs` is not None. I believe it works either way actually

File: src/core/domain_dataclasses.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+@dataclass
+class User:
+    """Class representing the user running the Pebble workload services."""
+
+    name: str
+    group: str
+
+
+@dataclass
+class S3ConnectionInfo:
```
Comment (welpaolo) [inline src/core/domain_dataclasses.py:20]: I believe this is just an alternative right?

**Review by welpaolo — APPROVED** [review]
Great work @deusebio :) I really like how things shape with this new pattern. 
I left some comments to better understand some items.

File: src/common/k8s.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    def read(self, path: str) -> list[str]:
+        """Read from a file.
+
+        Args:
+            path: the full filepath to be read
+
+        Returns:
+            content of the file
+        """
+        if not self.container.exists(path):
```
[the remark this answers is not in the document]
Comment (deusebio) [reply src/common/k8s.py:42]: Uhm, I agree. An empty file is different from a not existing file. However, I did it in this way as 'm kinda extending the Workload class, and some implementation where already provided in Kafka and ZooKeeper (see [here](https://github.com/canonical/kafka-operator/blob/main/src/workload.py#L58)). 

Should we prescriptive to have charms with same behaviour for this? Maybe I would say so (given it would be good to have consistent `read` and `write` implementations), but in this case I would give +1 to @welpaolo 

@marcoppenheimer @juditnovak

**Review by theoctober19th — APPROVED** [review]

File: src/core/domain_dataclasses.py (hunk 2 above)
Comment (deusebio) [reply src/core/domain_dataclasses.py:20]: Yes, it is the dataclass abstraction I tried. But I believe to make it possible we have to resolve not only the exposing the data but also providing a way to set data. I will remove this and investigate this at a later stage

File: src/events/base.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+# See LICENSE file for licensing details.
+
+"""Base utilities exposing common functionalities for all Events classes."""
+
+from functools import wraps
+from typing import Callable
+
+from ops import CharmBase, EventBase, Object, StatusBase
+
+from core.context import AuthProxyConfig, Context, IngressUrl, S3ConnectionInfo, Status
```
[the remark this answers is not in the document]
Comment (deusebio) [reply src/events/base.py:12]: Uhm, that's a small (but maybe relevant) point: here we could centralize external dependencies (charmlibs) into one module (e.g. core) instead of having the external imports on all modules. This has some, but not critical, advantages, like:

1. having a single proxy module where external dependencies are centralized 
2. Possibility of enhance external classes with extra capabilities, or patch behaviour. 

It is a pattern I kinda like, but up to you. I don't have strong opinion

File: src/common/workload.py
Comment (marcoppenheimer) [inline src/common/workload.py:1]: **question:** What is this file adding? To me it appears to be redundant, as the actual logic is in `core/workload.py` already?

I'm not sure I see the sense of spreading an API out across 4 files, with `k8s`, `workload`, `workload` and `workload`. Could they be coalesced?

File: src/common/k8s.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/common/k8s.py:42]: IMO whatever we choose to return is up to us.
Personally, I like empty lists as they're falsey, and avoids me having to `try: except` all the time, when really all I care about is if there is 'content' there or not, but I wouldn't want to impose that on other teams if they feel differently!

File: src/workload.py [hunk 5]
```diff
@@ -204,7 +88,20 @@ def ready(self) -> bool:
...
-        # We could use pebble health checks here
+        try:
+            service = self.container.get_service(self.HISTORY_SERVER_SERVICE)
+        except ops.pebble.ConnectionError:
+            self.logger.debug(f"Service {self.HISTORY_SERVER_SERVICE} not running")
+            return False
+
+        return service.is_running()
+
+    def set_environment(self, envs: dict[str, str | None]):
```
Comment (marcoppenheimer) [inline src/workload.py:101]: todo:
```
Pyright: Method "set_environment" overrides class "SparkHistoryWorkloadBase" in an incompatible manner
  Parameter 2 name mismatch: base parameter is named "env", override parameter is named "envs" [reportIncompatibleMethodOverride]
```

File: src/workload.py [hunk 6]
```diff
@@ -3,199 +3,83 @@
...
+from core.workload import HistoryServerPaths, SparkHistoryWorkloadBase
 
-    @abstractmethod
-    def exec(
-        self, command: str, env: dict[str, str] | None = None, working_dir: str | None = None
-    ) -> str:
-        """Runs a command on the workload substrate."""
-        ...
 
+class SparkHistoryServer(SparkHistoryWorkloadBase, K8sWorkload, WithLogging):
```
Comment (marcoppenheimer) [inline src/workload.py:17]: **question:** Related to my other comment. Why aren't we the implementing methods currently in `K8sWorkload` in `SparkHistoryWorkloadBase`? or better yet, just in plain `SparkHisotryServer`? What does the double inheritance here solve?

File: src/common/k8s.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply src/common/k8s.py:42]: My suggestion here would be to have a clean signature (that raises exceptions, given non-existence IS a different case of having a file empty) in the "general" class, that applies to all charms. If we don't want to be prescriptive, in the product specific implementation one can override this by wrapping the try except. 

Actually, I believe it is a good decorator use case to wrap try expect (with some default value to return when this happens), something like:

```
@safe(default=[])
def function_that_can_fail(....):
```

I believe I have implemented this previously in my life and I found it quite useful, for similar reason as Marc said. But I would not make it generic.

File: src/workload.py (hunk 6 above)
Comment (deusebio) [reply src/workload.py:17]: I noticed that in kafka and zookeeper we are implementing read/write operations in the final class. These are operations very common for which I would believe it is desirable to have common implementations, one for k8s and one for VM. I don't think the business logic there would change between products. Therefore:

* SparkHistoryWorkloadBase is an substrate independent class (where you specify product specific things, e.g. `Paths`, but that are substrate independent)
* K8sWorkload is a product independent class (where you specify substrate specific things, e.g. `read` and `write`)

When you add them together, you can get the most-complete base for the final class, which is substrate-specific and product specific.

Comment (deusebio) [reply src/workload.py:17]: I believe I created an image at a certain point on Excalidraw.io

File: src/workload.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (deusebio) [reply src/workload.py:44]: Done! This way we avoid one recursion. It is not much, but as you pointed out, not really needed, I agree.

File: src/workload.py (hunk 6 above)
Comment (marcoppenheimer) [reply src/workload.py:17]: I think there are two issues with it that I see from first glance.

1. The assumption that any implementation can be 'common'. I don't think that's the case. For charms based on Snaps, we want to `chown` every file on `write()`. That isn't the case for Deb charms. There will be lots of cases where behavior diverges slightly, and we shouldn't be restrictive for that for the sake of it
2. If we DO choose to have a shared implementation for the team, I think it's too early to say that what is presented in this PR is how it will look like. It seems to me like an early optimization to include it here now, and one that increases complexity without any clear upsides

tl;dr - for the current state of the spark charm, I don't think we need double-inheritance. Especially for a charm that only exists on K8s

Comment (deusebio) [reply src/workload.py:17]: TLDR I don't think using `common` and/or double inheritance creates a lot more complexity (actually provides structure to me), and although I agree it may not be needed by a simple charm (as this one) I see it as a step to foster code-sharing,  common functionalities and collaboration across team. But, here, up to @welpaolo and @theoctober19th . You guys will have to deal with the code more than me, so I don't want to create something that creates friction for you.

Longer reply with my thoughts :smile: 

1. I don't say "ANY" implementation should be in common but only charm-independent ones. We should try to factor out charm-independent functionalities and "read/write"  does seem such a case to me. it is fine to have slightly different behaviour that we can handle with function/constructors arguments, class attributes or different classes (e.g. K8s vs VM). This would provide the foundation for having more homogeneous and maintainable code bases. 

2. Maybe it is not needed in this specific context, but I'm trying to also use these PRs to test/suggest patterns, possibly pushing some ideas out. I wasn't planning to push this into a library and say this is what the code will look like, but I believe it would be good - as a collaboration pattern - if we centralize a `common` module (or another name is fine too :D) with general/common abstractions we see (or also extending existing classes) in **all projects**. It would make discussion easier and provide a first landing space for general functionalities, without needing to release/discuss/agree on general libraries (and be blocked as a feature gets developed) and testing them for some time to evaluate pros/cons. I have implemented this in the past (as a collaboration pattern between different teams with more specific goals), and it worked well

Comment (theoctober19th) [reply src/workload.py:17]: I don't have strong reasons to oppose the implementation. However, the only small thing I wanted to ask was, is `SparkHistoryServer extends SparkHistoryWorkloadBase extends K8sWorkload` not a considerable approach to this (since Spark History Server is a K8s only charm)? Basically, what if we use hierarchical inheritance instead of multiple inheritance?

Comment (welpaolo) [reply src/workload.py:17]: I don't have any blocker for the implementation that @deusebio proposed, I like the idea of having a common read/write functions. 
We can evaluate the points raised by @marcoppenheimer in a couple of pulses and revert this approach if it is not working

**Review by welpaolo — APPROVED** [review]

**Comment by deusebio** [comment]
Merging, given @theoctober19th and @welpaolo approvals, as this is really blocking the other PR #32 to be finalized. @juditnovak feel free to share thoughts (if any) anyway for future tuning and improvement. Thanks!

**Comment by juditnovak** [comment]
I was late to the party indeed :sweat_smile: 

Joke aside: though not direct comments, but we looked at the PR together with @marcoppenheimer so his comments were representing my opinon as well.

Having said that, I'd like to +1 on his considerations regarding common read/write interface or not.
I also prefer his approach within the guidelines of "Simple is better than complex".

No matter what, most importantly I guess we all agree on the following: we all believe this is typically the level of detail we should leave as the freedom of the developer of the charm... (I.e. not part of any conventional standards or strongly recommended patterns.)
